# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/TLSTransportLayer.cpp
# Commit: ed7b5e4aa348
# Full Hash: ed7b5e4aa348947a1822e380a8fb561306a9ad8b
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2023-08-16 03:43:43
# Regressor Bug: 1847904
# File Overlap Count: 1
# Description:
#   Bug 1847904 - Dispatch PR_Close to socketthread in ~TLSTransportLayer r=keeler,necko-reviewers,kershaw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D185978
# ==============================================================================

diff -r e20f92633534 -r ed7b5e4aa348 netwerk/protocol/http/TLSTransportLayer.cpp
--- a/netwerk/protocol/http/TLSTransportLayer.cpp	Tue Aug 15 11:59:00 2023 +0300
+++ b/netwerk/protocol/http/TLSTransportLayer.cpp	Tue Aug 15 09:58:03 2023 +0000
@@ -342,13 +342,28 @@
 }
 
 TLSTransportLayer::~TLSTransportLayer() {
-  MOZ_ASSERT(OnSocketThread(), "not on socket thread");
   LOG(("TLSTransportLayer dtor this=[%p]", this));
+
+  auto closeSocket = [fd = mFD, socketControl = std::move(mTLSSocketControl)] {
+    MOZ_ASSERT(OnSocketThread(), "not on socket thread");
+    if (fd) {
+      PR_Close(fd);
+    }
+    // mTLSSocketControl must be released after closing fd
+  };
+
+  if (OnSocketThread()) {
+    closeSocket();
+    return;
+  }
+
   if (mFD) {
-    PR_Close(mFD);
-    mFD = nullptr;
+    gSocketTransportService->Dispatch(NS_NewRunnableFunction(
+        "TLSTransportLayer::~TLSTransportLayer", std::move(closeSocket)));
+    return;
   }
-  mTLSSocketControl = nullptr;
+
+  // mTLSSocketControl is released when closeSocket goes out of scope
 }
 
 bool TLSTransportLayer::Init(const char* aTLSHost, int32_t aTLSPort) {
